astropy/units/quantity_helper/function_helpers.py

```diff
@@ -1103,3 +1112,120 @@ def unstructured_to_structured(arr, dtype, *args, **kwargs):
 
     return (arr.to_value(arr.unit), dtype) + args, kwargs, target_unit, None
+
+
+def _izip_units_flat(iterable: Iterable[StructuredUnit | Unit] | StructuredUnit) -> Iterator:
+    """Returns an iterator of collapsing any nested unit structure.
+
+    Parameters
+    ----------
+    iterable : Iterable[StructuredUnit | Unit] or StructuredUnit
+        A structured unit or iterable thereof.
+
+    Yields
+    ------
+    (str, unit)
+    """
+    from astropy.units import StructuredUnit
+
+    # Make Structured unit.
+    units = iterable if isinstance(iterable, StructuredUnit) else StructuredUnit(iterable)
+
+    # Yield from structured unit.
+    for k, v in units.items():
+        if isinstance(v, StructuredUnit):
+            yield from _izip_units_flat(v)
+        else:
+            yield (k, v)
+
+
+def merge_units(*units: UnitBase | StructuredUnit | None) -> StructuredUnit:
+    """Merge a sequence of units into a structured unit.
+
+    .. todo::
+
+        Either make this function public or incorporate into ``StructuredUnit``.
+
+    Parameters
+    ----------
+    *units : `astropy.units.UnitBase` or `astropy.units.StructuredUnit`
+        The units to merge into one structured unit.
+
+    Returns
+    -------
+    `astropy.units.StructuredUnit`
+        Structured unit of constituent unit.
+    """
+    from astropy.units import StructuredUnit, UnitBase
+
+    # filter units, excluding None
+    actual_units = tuple(unit for unit in units if isinstance(unit, (UnitBase, StructuredUnit)))
+    flat = tuple(_izip_units_flat(StructuredUnit(actual_units)))
+    return StructuredUnit(tuple(uu for _, uu in flat), names=tuple(n for n, _ in flat))
+
+
+@function_helper(helps=rfn.merge_arrays)
+def merge_arrays(
+    seqarrays: Quantity | Sequence[Quantity],
+    fill_value: float = -1,
+    flatten: bool = False,
+    usemask: bool = False,
+    asrecarray: bool = False,
+) -> tuple[tuple[Any, ...], dict[str, Any], StructuredUnit, None]:
+    """Merge structured Quantities field by field.
+
+    Parameters
+    ----------
+    seqarrays : sequence of Quantity
+        Sequence of arrays.
+    fill_value : {float}, optional
+        Filling value used to pad missing data on the shorter arrays.
+    flatten : {False, True}, optional
+        Whether to collapse nested fields.
+    usemask : {False, True}, optional
+        Whether to return a masked array or not.
+
+        .. warning::
+            Not yet supported.
+    asrecarray : {False, True}, optional
+        Whether to return a recarray (MaskedRecords) or not.
+
+        .. warning::
+            Not yet supported.
+
+    Returns
+    -------
+    Quantity
+        Merged structured Quantity.
+
+    Raises
+    ------
+    ValueError
+        If ``asrecarray`` is `True`.
+        If ``usemask`` is `True`.
+    """
+    from astropy.units import Quantity, StructuredUnit
+
+    if asrecarray:
+        raise ValueError("asrecarray=True is not supported.")
+    if usemask:
```

Comment:
Maybe add a `TODO: use MaskedQuantity for this case.`

Reply:
☑